### PR TITLE
Segment out SymbolResolver to a trait

### PR DESCRIPTION
**Pull request by addisoncrump**
Makes symbol resolution a trait. Begins to address #705 and extend #339, allows users to define custom implementations in the meantime.

### CODE REVIEW

**Comment by addisoncrump** [comment]
The public API changes should not affect any downstream users.

File: aya/src/maps/stack_trace.rs [hunk 1]
```diff
@@ -159,6 +159,21 @@ impl<'a, T: Borrow<MapData>> IntoIterator for &'a StackTraceMap<T> {
     }
 }
 
+/// A resolver for symbols based on an address obtained from a stacktrace.
```
Comment (tamird) [inline aya/src/maps/stack_trace.rs:162]: nit: "stack trace"?

File: aya/src/maps/stack_trace.rs [hunk 2]
```diff
@@ -159,6 +159,21 @@ impl<'a, T: Borrow<MapData>> IntoIterator for &'a StackTraceMap<T> {
...
+pub trait SymbolResolver {
+    /// Resolve a symbol for a given address, if possible.
+    fn resolve_sym(&self, addr: u64) -> Option<String>;
```
Comment (tamird) [inline aya/src/maps/stack_trace.rs:165]: Should this return `Option<&str>` or perhaps `Option<Cow<'_, str>>`?

File: aya/src/maps/stack_trace.rs [hunk 3]
```diff
@@ -159,6 +159,21 @@ impl<'a, T: Borrow<MapData>> IntoIterator for &'a StackTraceMap<T> {
...
+}
+
+/// The simplest resolver: a direct map from addresses to strings.
+pub type SimpleSymbolResolver = BTreeMap<u64, String>;
```
Comment (tamird) [inline aya/src/maps/stack_trace.rs:169]: should this implementation be in util.rs? It'd be nice if it lived next to where it's constructed.

**Review by tamird — COMMENTED** [review]
You'll need to run `cargo xtask public-api` to regenerate the public API fixture.

File: aya/src/maps/stack_trace.rs [hunk 4]
```diff
@@ -159,6 +159,12 @@ impl<'a, T: Borrow<MapData>> IntoIterator for &'a StackTraceMap<T> {
     }
 }
 
+/// A resolver for symbols based on an address obtained from a stack trace.
+pub trait SymbolResolver {
+    /// Resolve a symbol for a given address, if possible.
+    fn resolve_sym(&self, addr: u64) -> Option<&str>;
```
Comment (alessandrod) [inline aya/src/maps/stack_trace.rs:165]: nit `resolve_symbol`? Trait being symbol method being sym triggers me 😅

File: aya/src/util.rs [hunk 5]
```diff
@@ -14,6 +14,7 @@ use crate::{
     Pod,
 };
 
+use crate::maps::stack_trace::SymbolResolver;
```
Comment (alessandrod) [inline aya/src/util.rs:17]: fold this in the `use crate` above

File: aya/src/maps/stack_trace.rs (hunk 2 above)
Comment (alessandrod) [reply aya/src/maps/stack_trace.rs]: Yeah probably Cow

Comment (addisoncrump) [reply aya/src/maps/stack_trace.rs]: Didn't get notified for this, whoops. Should now be fixed.

File: aya/src/util.rs [hunk 6]
```diff
@@ -206,8 +207,8 @@ fn parse_cpu_ranges(data: &str) -> Result<Vec<u32>, ()> {
 pub type SimpleSymbolResolver = BTreeMap<u64, String>;
 
 impl SymbolResolver for SimpleSymbolResolver {
-    fn resolve_symbol(&self, addr: u64) -> Option<&str> {
-        self.range(..=addr).next_back().map(|(_, s)| s.as_str())
+    fn resolve_symbol(&self, addr: u64) -> Option<Cow<'_, str>> {
+        self.range(..=addr).next_back().map(|(_, s)| Cow::from(s))
```
Comment (tamird) [inline aya/src/util.rs:211]: nit: can be s.into(), I think

**Comment by tamird** [comment]
LGTM. @dave-tucker should have a look. Please squash the commits.

File: aya/src/maps/stack_trace.rs [hunk 7]
```diff
@@ -174,12 +185,9 @@ impl StackTrace {
...
     /// user-space traces you need to provide the symbols, for example loading
     /// them from debug info.
-    pub fn resolve(&mut self, symbols: &BTreeMap<u64, String>) -> &StackTrace {
+    pub fn resolve<R: SymbolResolver>(&mut self, symbols: &R) -> &StackTrace {
         for frame in self.frames.iter_mut() {
-            frame.symbol_name = symbols
-                .range(..=frame.ip)
-                .next_back()
-                .map(|(_, s)| s.clone())
+            frame.symbol_name = symbols.resolve_symbol(frame.ip).map(|s| s.to_string())
```
Comment (tamird) [inline aya/src/maps/stack_trace.rs:190]: s/to_string/to_owned/

Comment (tamird) [reply aya/src/maps/stack_trace.rs:190]: I don't think I understand; `str::to_owned` returns `String`. https://doc.rust-lang.org/stable/src/alloc/str.rs.html#205

Comment (addisoncrump) [reply aya/src/maps/stack_trace.rs:190]:
```
[addisoncrump@fuzzbox aya]$ cargo build
   Compiling aya v0.11.0 (/home/addisoncrump/git/aya/aya)
error[E0308]: mismatched types
   --> aya/src/maps/stack_trace.rs:190:33
    |
190 |             frame.symbol_name = symbols.resolve_symbol(frame.ip).map(|s| s.to_owned())
    |             -----------------   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Option<String>`, found `Option<Cow<'_, str>>`
    |             |
    |             expected due to the type of this binding
    |
    = note: expected enum `Option<String>`
               found enum `Option<Cow<'_, str>>`
```

Comment (tamird) [reply aya/src/maps/stack_trace.rs:190]: Ah, it's [`into_owned`](https://doc.rust-lang.org/stable/std/borrow/enum.Cow.html#method.into_owned), not `to_owned`.

**Comment by addisoncrump** [comment]
Build failure due to an upstream crate. That version is now yanked and this should work again.

File: aya/src/maps/stack_trace.rs (hunk 7 above)
Comment (addisoncrump) [reply aya/src/maps/stack_trace.rs]: Resolved

**Review by dave-tucker — APPROVED** [review]

**Pull request merged**